hal/watchdog_api.h

```diff
@@ -0,0 +1,156 @@
+/** \addtogroup hal */
+/** @{*/
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_WATCHDOG_API_H
+#define MBED_WATCHDOG_API_H
+
+#include <stdbool.h>
+#include <stdint.h>
+
+typedef struct
+{
+  /**
+  * Refresh value for the watchdog in milliseconds. The maximum value of this
+  * setting is platform dependent, to find the maximum value for the current
+  * platform call hal_watchdog_get_max_timeout(void)
+  */
+  uint32_t timeout_ms;
+  /**
+   * Configures the watchdog for windowed mode of operation instead of running
+   * the independent watchdog. In this mode the a restriction is placed on the
+   * the time period in which the watchdog can be refreshed/kicked. If the
+   * watchdog is kicked too soon after it has last been refreshed the system
+   * will be reset. The period of time in which the reset will be triggered is
+   * defined by window_ms. This value is false by default.
+   */
+  bool enable_window;
+  /**
+   * Specifies the time window for the watchdog window in milliseconds. If the
+   * watchdog is configured to run in windowed mode kicking the watchdog less
+   * than this many milliseconds after it has last been kicked will trigger a
+   * system reset. This value must be less than timeout_ms.
+   */
+  uint32_t window_ms;
+  /**
+   * Configures the watchdog timer to run while the core is in sleep mode. By
+   * default when the system is put into the sleep the watchdog timer is paused.
+   * Enabling this setting causes the timer to countdown during this time. This
+   * flag is disabled by default.
+   */
+  bool enable_sleep;
```

Comment:
Is that supported on all the platforms? Also is watchdog timer paused during sleep on all the platforms?

Reply:
It's probably not supported on all platforms. It is supported by STM32, NORDIC, and K64F family boards though. I don't recall which the default option is on each but it is configurable on each of the boards.

Reply:
I would have expected the default behavior to be: don't pause timer when sleeping, so that the MCU wont be stuck in sleep forever.

Reply:
That would make more sense and seems to be the default for all 3 boards I looked at.

Reply:
What does it make default? If someone passes in this structure it's alwasy true or false one way or another.

Reply:
> If someone passes in this structure it's alwasy true or false one way or another.

True but it's worth pointing out that such thing as a default config exists (at least for **K64F**: https://github.com/ARMmbed/mbed-os/blob/4e222952d757e378d5a591ca7fd143dcda24fc7e/targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/drivers/fsl_wdog.h#L187) and might be useful as show here: https://github.com/ARMmbed/mbed-os/blob/4e222952d757e378d5a591ca7fd143dcda24fc7e/targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/TARGET_MCU_K64F/drivers/fsl_wdog.h#L198-L202

Regarding this new HAL API you may consider providing a similar function i.e. `void hal_watchdog_get_default_config(watchdog_config_t *config);`

Reply:
Let's not model our APIs on KSDK, we dont really need it our driver can handle the configuration.